### PR TITLE
Draft of initial RMC documentation

### PR DESCRIPTION
**Pull request by tedinski**
### Description of changes: 

Some initial documentation for RMC.

### Resolved issues:

N/A

### Call-outs:

Nothing special.

### Testing:

* How is this change tested? -

* Is this a refactor change? -

### Checklist
- [x] Each commit message has a non-empty body, explaining why the change was made
- n/a Methods or procedures are documented
- n/a Regression or unit tests are included, or existing tests cover the modified code
- [x] My PR is restricted to a single feature or bugfix

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 and MIT licenses.

### CODE REVIEW

File: rmc-docs/src/getting-started.md [hunk 1]
```diff
@@ -1,9 +1,17 @@
...
+RMC is especially useful for verifying unsafe code in Rust, where many of the language's usual guarantees can no longer be checked by the compiler.
 
-```rust
-fn main() {
-    assert!(1 == 1);
-}
-```
+## Project Status
+
+RMC is currently in the initial development phase, and has not yet reached even a "developer preview" release state.
```
Comment (nchong-at-aws) [inline rmc-docs/src/getting-started.md:9]: Remove 'and has not yet reached even a "developer preview" release state.' since we aren't committing to a time for that.

File: rmc-docs/src/getting-started.md [hunk 2]
```diff
@@ -1,9 +1,17 @@
...
+It is under active development, but it does not yet support all rust language features.
```
Comment (nchong-at-aws) [inline rmc-docs/src/getting-started.md:10]: Link to dashboard when it's ready!

File: rmc-docs/src/install-guide.md [hunk 3]
```diff
@@ -0,0 +1,85 @@
+# RMC Installation Guide
+
+RMC has not yet reached the point where releases are available, and so to use RMC you must currently build from source.
```
Comment (nchong-at-aws) [inline rmc-docs/src/install-guide.md:3]: Just asy "RMC must currently be built from source"

File: rmc-docs/src/install-guide.md [hunk 4]
```diff
@@ -0,0 +1,85 @@
...
+
+```
+./configure \
+    --enable-debug \
+    --set=llvm.download-ci-llvm=true \
+    --set=rust.debug-assertions-std=false \
+    --set=rust.deny-warnings=false
+```
+
+**NOTE: If you skip the above, builds may take a long time as all of LLVM would need to be built from scratch.**
```
Comment (nchong-at-aws) [inline rmc-docs/src/install-guide.md:36]: Be more specific and call out it's llvm.download-ci-llvm that's important?

File: rmc-docs/src/install-guide.md [hunk 5]
```diff
@@ -0,0 +1,85 @@
...
+
+Then build RMC:
+
+```
+./x.py build -i --stage 1 library/std
+```
+
+Then, optionally, run the regression tests:
```
Comment (nchong-at-aws) [inline rmc-docs/src/install-guide.md:44]: Say what the expected result is?

File: rmc-docs/src/rmc-tutorial.md [hunk 6]
```diff
@@ -0,0 +1,9 @@
+# RMC tutorial
+
+This tutorial will step you through a reasonable progression from simple to moderately complex tasks with RMC.
```
Comment (nchong-at-aws) [inline rmc-docs/src/rmc-tutorial.md:3]: Remove 'reasonable'. Maybe start with some motivation. Why do you want to invest time learning RMC?

File: rmc-docs/src/tool-comparison.md [hunk 7]
```diff
@@ -0,0 +1,20 @@
+# Comparison with other tools
```
Comment (nchong-at-aws) [inline rmc-docs/src/tool-comparison.md:1]: You could talk about the progression of: concrete inputs with testing, probabilistic inputs with fuzzing and symbolic inputs with model checking. This section has a few terms that need explaining: SMT, NP-complete.

File: rmc-docs/src/getting-started.md [hunk 8]
```diff
@@ -1,9 +1,17 @@
 # Getting started with RMC
 
-Hello, World!
+RMC is a Rust verification tool based on _model checking_.
+With RMC, you can ensure that broad classes of problems are absent from your Rust code by writing _proof harnesses_, which are broadly similar to tests (especially property tests.)
+RMC is especially useful for verifying unsafe code in Rust, where many of the language's usual guarantees can no longer be checked by the compiler.
```
Comment (avanhatt) [inline rmc-docs/src/getting-started.md:5]: Maybe add that RMC can also find panics in safe Rust code (such as slice indexing errors) and user-added assertions.

File: rmc-docs/src/getting-started.md (hunk 2 above)
Comment (avanhatt) [inline rmc-docs/src/getting-started.md:10]: nit: `Rust`

File: rmc-docs/src/install-guide.md [hunk 9]
```diff
@@ -0,0 +1,85 @@
+# RMC Installation Guide
+
+RMC has not yet reached the point where releases are available, and so to use RMC you must currently build from source.
+
+In general, the following dependencies are required:
+
+1. The dependencies needed to built `rustc`. RMC is a fork of the Rust compiler, and so we have the same minimum requirements.
```
Comment (avanhatt) [inline rmc-docs/src/install-guide.md:7]: Maybe add that `cargo` will handle most of the installation for these dependencies?

File: rmc-docs/src/install-guide.md [hunk 10]
```diff
@@ -0,0 +1,85 @@
...
+2. CBMC (>= 5.30.1)
```
Comment (avanhatt) [inline rmc-docs/src/install-guide.md:8]: Need links for CBMC and CBMC viewer

File: rmc-docs/src/install-guide.md [hunk 11]
```diff
@@ -0,0 +1,85 @@
...
+3. CBMC Viewer (>= 2.6)
+
+## Installing on Ubuntu 20.04
+
+We recommend trying out RMC with Ubuntu 20.04.
```
Comment (avanhatt) [inline rmc-docs/src/install-guide.md:13]: Add that we have also tested on MacOS?

File: rmc-docs/src/install-guide.md [hunk 12]
```diff
@@ -0,0 +1,85 @@
...
+
+Then, optionally, run the regression tests:
+
+```
+./scripts/rmc-regression.sh
+```
+
+## Try running RMC
+
+Get the RMC script in your path:
```
Comment (avanhatt) [inline rmc-docs/src/install-guide.md:52]: Recommend adding to `.bashrc`/`.bash_profile`?

File: rmc-docs/src/tool-comparison.md [hunk 13]
```diff
@@ -0,0 +1,20 @@
...
+
+**Fuzzing** (for example, with [`cargo-fuzz`](https://github.com/rust-fuzz/cargo-fuzz)) is a blind approach to random testing.
```
Comment (avanhatt) [inline rmc-docs/src/tool-comparison.md:3]: Could we avoid the word "blind" here? "General", "broad approach to automating random testing"?

File: rmc-docs/src/tool-comparison.md [hunk 14]
```diff
@@ -0,0 +1,20 @@
...
+A fuzzer generally provides an input of random bytes, and then examines fairly generic properties ("doesn't crash or commit undefined behavior") about the behavior of the resulting program.
+Fuzzers generally get their power through a kind of evolutionary algorithm that rewards new mutant inputs that "discover" new branches of the program under test.
+Fuzzers are excellent for testing security boundaries, precisely because they make no validity assumptions about the input.
+
+**Property testing** (for example, with [`proptest`](https://github.com/AltSysrq/proptest)) is a non-blind approach to random testing.
```
Comment (avanhatt) [inline rmc-docs/src/tool-comparison.md:8]: "more involved"/"higher effort" instead of non-blind?

File: rmc-docs/src/tool-comparison.md [hunk 15]
```diff
@@ -0,0 +1,20 @@
...
+Specific ranges of random values for specific types are generated several times, and the assertions of the test are checked for each.
+Tests in this style do actually state properties: "forall values (of some kind), this condition holds."
+But property testing can only sample randomly a few of those values to test (though property testing libraries frequently give interesting "edge cases" higher probability, making them more effective at bug-finding).
+
+**Model checking** is similar to these techniques in how you use them, but model checking is non-random and exhaustive.
```
Comment (avanhatt) [inline rmc-docs/src/tool-comparison.md:13]: "exhaustive (at least up to some bound on input or problem size)"

File: rmc-docs/src/tool-comparison.md [hunk 16]
```diff
@@ -0,0 +1,20 @@
...
+Thus, properties checked with a model checker are effectively proofs.
+Instead of naively trying all possible inputs (which could be infeasible and blow up exponentially), model checkers like RMC will cleverly encode program traces as "SMT" problems, and hand them off to SMT solvers.
```
Comment (avanhatt) [inline rmc-docs/src/tool-comparison.md:15]: RMC uses SAT, not SMT, no?

File: rmc-docs/src/install-guide.md [hunk 17]
```diff
@@ -0,0 +1,85 @@
...
+
+## Installing on Ubuntu 20.04
+
+We recommend trying out RMC with Ubuntu 20.04.
+The simplest way to install is following our CI scripts:
+
+```
+git clone git@github.com:model-checking/rmc.git
+cd rmc
+git submodule update --init
```
Comment (zhassan-aws) [inline rmc-docs/src/install-guide.md:19]: Is this command necessary (I don't recall running it in my initial setup). I just did that, and it's been running for over 30 minutes for me and seems to be cloning a lot of repositories:
```
Cloning into '/home/zyadh/rmc/firecracker'...
Cloning into '/home/zyadh/rmc/library/backtrace'...
Cloning into '/home/zyadh/rmc/library/stdarch'...
Cloning into '/home/zyadh/rmc/src/doc/book'...
Cloning into '/home/zyadh/rmc/src/doc/edition-guide'...
Cloning into '/home/zyadh/rmc/src/doc/embedded-book'...
Cloning into '/home/zyadh/rmc/src/doc/nomicon'...
Cloning into '/home/zyadh/rmc/src/doc/reference'...
Cloning into '/home/zyadh/rmc/src/doc/rust-by-example'...
Cloning into '/home/zyadh/rmc/src/doc/rustc-dev-guide'...
Cloning into '/home/zyadh/rmc/src/llvm-project'...
```

File: rmc-docs/src/rmc-tutorial.md (hunk 6 above)
Comment (tedinski) [reply rmc-docs/src/rmc-tutorial.md]: Did the first. I also added more context.

File: rmc-docs/src/install-guide.md (hunk 17 above)
Comment (tedinski) [reply rmc-docs/src/install-guide.md:19]: I've assumed it was necessary, since our CI is doing it.

**Review by avanhatt — APPROVED** [review]
LGTM!
